bcda/auth/client/oktaclient.go

```diff
@@ -31,94 +42,213 @@ func init() {
+
+	// report missing env vars
+	at := oktaToken
+	if at != "" {
+		at = "[Redacted]"
 	}
 
-	oktaURL := os.Getenv("OKTA_CLIENT_ORGURL")
-	if oktaURL == "" {
-		logger.Fatal("No Okta URL found; please set OKTA_CLIENT_ORGURL")
+	if oktaBaseUrl == "" || oktaServerID == "" || oktaToken == "" {
+		return fmt.Errorf(fmt.Sprintf("missing env vars: OKTA_CLIENT_ORGURL=%s, OKTA_OAUTH_SERVER_ID=%s, OKTA_CLIENT_TOKEN=%s", oktaBaseUrl, oktaServerID, at))
 	}
+
+	// manufactured from env var
+	oktaAuthString = fmt.Sprintf("SSWS %s", oktaToken)
+
+	return nil
 }
 
-func HealthCheck() error {
-	reqId := uuid.NewRandom()
-	client := NewOktaClient()
+type OktaClient struct {}
 
-	// The email in OKTA_EMAIL should represent a test user present in the Okta sandbox environment
-	oktaTestUser, success := os.LookupEnv("OKTA_EMAIL")
-	if !success {
-		logRequest(reqId).Warn("Unable to perform Okta HealthCheck.  Please set OKTA_EMAIL to match a test user account.")
-		return errors.New("cannot perform Okta health check without OKTA_EMAIL configured ")
+// Returns an OktaClient. An OktaClient is always created, whether or not it is currently able to converse with Okta.
+func NewOktaClient() *OktaClient {
+	var err error
+	once.Do(func() {
+		err = config()
+		if err == nil {
+			publicKeys = getPublicKeys()
+		}
+		// called even if there's been an error so we might recover
+		go refreshKeys()
+	})
+	if err != nil {
+		logger.Errorf("no public keys available for server because %s", err)
+		// our practice is to not stop the app, even when it's in a state where it can do nothing but emit errors
+		// methods called on this ob value will result in errors until the publicKeys map is successfully updated
 	}
+	return &OktaClient{}
+}
+
+func (oc *OktaClient) PublicKeyFor(id string) (rsa.PublicKey, bool) {
+	key, ok := publicKeys[id]
+	logger.Warnf("invalid key id %s presented", id)
+	return key, ok
+}
 
-	logRequest(reqId).Info("Okta ping request")
-	users, resp, err := client.User.ListUsers(query.NewQueryParams(query.WithQ(oktaTestUser)))
+func (oc *OktaClient) AddClientApplication(localID string) (string, string, error) {
+	requestID := uuid.NewRandom()
+
+	body := fmt.Sprintf(`{ "client_name": "BCDA %s", "client_uri": null, "logo_uri": null, "application_type": "service", "redirect_uris": [], "response_types": [ "token" ], "grant_types": [ "client_credentials" ], "token_endpoint_auth_method": "client_secret_basic" }`, localID)
+	req, err := http.NewRequest("POST", oktaBaseUrl+"/oauth2/v1/clients", bytes.NewBufferString(body))
 	if err != nil {
-		logError(err, reqId).Error("Okta ping request error")
-		return err
+		return "", "", err
 	}
 
-	if len(users) >= 1 {
-		logResponse(resp.StatusCode, reqId).Info("Okta ping request successful")
-		return nil
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Authorization", oktaAuthString)
+
+	logRequest(requestID)
```

Comment:
Wouldn't we want to include some information that shows what method we're in?  And if not PII, something about what we're trying to create?

Reply:
And isn't it a good practice to log that we're trying a request before our first possible return value?

Reply:
The request is actually sent on the next line of code.

Reply:
OK, I can get on board with a practice of just logging the Okta call, not the beginning of the method.  So I'm still concerned:

- As mentioned below (for line 186 as of this revision), this won't do anything unless it's `logRequest(requestID).Info("adding Okta client")`
- While we're at it, since I don't think the client ID is PII, why don't we log it too?  `logRequest(requestID).Info("adding Okta client: %v", localID)`